src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java

```diff
@@ -719,45 +716,50 @@ public void visitCtDo(CtDo doLoop) {
 
 	@Override
 	public <T> void visitCtThisAccess(CtThisAccess<T> thisAccess) {
-		enterCtExpression(thisAccess);
-		if (thisAccess.getTarget() != null && thisAccess.getTarget() instanceof CtTypeAccess
-				&& !tryToInitializeFinalFieldInConstructor(thisAccess)
-				&& !thisAccess.isImplicit()
-				&& !thisAccess.getTarget().isImplicit()) {
-			final CtTypeReference accessedType = ((CtTypeAccess) thisAccess.getTarget()).getAccessedType();
-			if (accessedType.isLocalType()) {
-				printer.write(accessedType.getSimpleName().replaceAll("^[0-9]*", "") + ".");
-			} else if (!accessedType.isAnonymous()) {
-				visitCtTypeReferenceWithoutGenerics(accessedType);
-				printer.write(".");
+		try {
+			enterCtExpression(thisAccess);
+
+			// we only write qualified this when this is required
+			// this is good both in fully-qualified mode and in readable (with-imports) mode
+			// the implicit information is used for analysis (eg are visibility caused by implicit bugs?) but
+			// not for pretty-printing
+			CtTypeReference thisTarget = ((CtTypeAccess) thisAccess.getTarget()).getAccessedType();
+
+			// the simplest case: we always print "this" if we're in the top-level class,
+			// this is shorter (no qualified this), explicit, and less fragile wrt transformation
+			if (thisTarget == null || (thisAccess.getParent(CtType.class) != null && thisAccess.getParent(CtType.class).isTopLevel())) {
+				printer.write("this");
+				return; // still go through finally block below
 			}
-		}
-		if (!thisAccess.isImplicit()) {
-			printer.write("this");
-		}
-		exitCtExpression(thisAccess);
-	}
 
-	/**
-	 * Check if the this access expression is a target of a private final field in a constructor.
-	 */
-	private <T> boolean tryToInitializeFinalFieldInConstructor(CtThisAccess<T> thisAccess) {
-		try {
-			final CtElement parent = thisAccess.getParent();
-			if (!(parent instanceof CtFieldWrite) || !thisAccess.equals(((CtFieldWrite) parent).getTarget()) || thisAccess.getParent(CtConstructor.class) == null) {
-				return false;
+			// TODO find a way to remove this check
+			// ideally we sould not rely on implicitness
+			// for correct pretty-printing, because this is fragile
+			// wrt to bugs in the model and when refactoring
+			if (thisTarget.isImplicit()) {
+				// write nothing, "this" is implicit and we unfortunately cannot always know
+				// what the good target is in JDTTreeBuilder
+				return;
```

Comment:
no because the caller checks for implicitness as well (see ctVisitInvocation for instance).

however, this is an interesting question: who is responsible for printing the "."? probably the parent element, as done today (but consistently?).

however instead of 
```
if (target.isImplicit()) {
  scan(target);
  print(".")
}
```
I would prefer 
```
scan(target);
if (somethingHasBeenWritten()) {
  print(".")
}
```
This would be less fragile.

(outside the scope of this PR but I'd like to have your opinion on this)
```